File: shared/utils/mcp_utils.py
#!/usr/bin/env python

# SPDX-FileCopyrightText: 2025 Weibo, Inc.
#
# SPDX-License-Identifier: Apache-2.0

# -*- coding: utf-8 -*-

from __future__ import annotations

import re
from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple

from shared.logger import setup_logger

if TYPE_CHECKING:
    from shared.models.execution import ExecutionRequest

logger = setup_logger("mcp_utils")


def extract_mcp_servers_config(config: Dict[str, Any]) -> Optional[Any]:
    """
    Extract MCP servers configuration supporting multiple formats.
    Priority order:
    1) {"mcpServers": {"mcpServers": {...}}}
       {"mcpServers": {"mcp_servers": {...}}}
    2) {"mcp_servers": {"mcp_servers": {...}}}
       {"mcp_servers": {"mcpServers": {...}}}
    3) {"mcpServers": {...}}
    4) {"mcp_servers": {...}}
    5) Fallback: config.get("mcp_servers")
    """
    try:
        # Collect candidates as tuples: (variant, outer_key, inner_key, cfg)
        # variant: "double" or "single"
        candidates: List[Tuple[str, str, str, Dict[str, Any]]] = []
        for outer_key in ("mcpServers", "mcp_servers"):
            val = config.get(outer_key)
            if isinstance(val, dict):
                found_double = False
                # Support cross-key inner nesting:
                for inner_key in ("mcpServers", "mcp_servers"):
                    nested = val.get(inner_key)
                    if isinstance(nested, dict) and nested:
                        logger.info(
                            f"Detected double-nested MCP config outer='{outer_key}', inner='{inner_key}'."
                        )
                        candidates.append(("double", outer_key, inner_key, nested))
                        found_double = True
                # If no valid double-nested dicts were found, treat the dict itself as single
                if not found_double and val:
                    logger.info(
                        f"Detected single-nested MCP config under '{outer_key}'."
                    )
                    candidates.append(("single", outer_key, "", val))

        # Priority selection (strict order)
        # 1) {"mcpServers": {"mcpServers": {...}}}
        for variant, outer, inner, cfg in candidates:
            if variant == "double" and outer == "mcpServers" and inner == "mcpServers":
                return cfg
        #    {"mcpServers": {"mcp_servers": {...}}}
        for variant, outer, inner, cfg in candidates:
            if variant == "double" and outer == "mcpServers" and inner == "mcp_servers":
                return cfg
        # 2) {"mcp_servers": {"mcp_servers": {...}}}
        for variant, outer, inner, cfg in candidates:
            if (
                variant == "double"
                and outer == "mcp_servers"
                and inner == "mcp_servers"
            ):
                return cfg
        #    {"mcp_servers": {"mcpServers": {...}}}
        for variant, outer, inner, cfg in candidates:
            if variant == "double" and outer == "mcp_servers" and inner == "mcpServers":
                return cfg
        # 3) {"mcpServers": {...}}
        for variant, outer, inner, cfg in candidates:
            if variant == "single" and outer == "mcpServers":
                return cfg
        # 4) {"mcp_servers": {...}}
        for variant, outer, inner, cfg in candidates:
            if variant == "single" and outer == "mcp_servers":
                return cfg

        # 5) Fallback: original behavior
        fallback = config.get("mcp_servers")
        if fallback is not None:
            logger.info("Using fallback 'mcp_servers' configuration.")
            return fallback
    except Exception as e:
        logger.warning(f"Failed to extract MCP servers configuration: {str(e)}")
    return None


def _get_nested_value(obj: Optional["ExecutionRequest"], path: str) -> Optional[Any]:
    """
    Get a value from an ExecutionRequest object using dot-separated path.
    Supports object attribute access and list index access.

    Args:
        obj: The ExecutionRequest object to search in
        path: Dot-separated path string, e.g., "user.name", "git_repo", or "bot.0.name"

    Returns:
        The value at the path, or None if the path doesn't exist

    Examples:
        >>> from shared.models.execution import ExecutionRequest
        >>> obj = ExecutionRequest(user={"name": "John", "id": 123})
        >>> _get_nested_value(obj, "user.name")
        'John'
        >>> _get_nested_value(obj, "user.email")
        None
    """
    if not obj or not path:
        return None

    keys = path.split(".")
    current: Any = obj

    for key in keys:
        if isinstance(current, list):
            # Try to parse key as integer index
            try:
                index = int(key)
                if 0 <= index < len(current):
                    current = current[index]
                else:
                    return None
            except ValueError:
                # Key is not a valid integer
                return None
        elif isinstance(current, dict):
            # Dict key access (for nested dicts like user, bot items)
            # Only access dict key if it exists, otherwise return None
            if key in current:
                current = current[key]
            else:
                return None
        elif hasattr(current, key):
            # Object attribute access (for ExecutionRequest and nested objects)
            current = getattr(current, key)
        else:
            return None

    return current


def _replace_placeholders_in_string(
    text: str, task_data: Optional["ExecutionRequest"]
) -> str:
    """
    Replace all ${{path}} placeholders in a string with values from task_data.

    Args:
        text: The string containing placeholders
        task_data: Data source for replacement values (ExecutionRequest object).

    Returns:
        The string with placeholders replaced. If a path doesn't exist in task_data,
        the original placeholder is preserved.

    Examples:
        >>> from shared.models.execution import ExecutionRequest
        >>> task_data = ExecutionRequest(user={"name": "John"}, git_repo="owner/repo")
        >>> _replace_placeholders_in_string("Hello ${{user.name}}", task_data)
        'Hello John'
        >>> _replace_placeholders_in_string("${{user.email}}", task_data)
        '${{user.email}}'
    """
    # Pattern to match ${{path.to.value}}
    pattern = r"\$\{\{([^}]+)\}\}"

    def replace_match(match: re.Match) -> str:
        path = match.group(1).strip()
        value = _get_nested_value(task_data, path)
        if value is not None:
            # Convert non-string values to string
            return str(value)
        else:
            # Keep original placeholder if path not found
            return match.group(0)

    return re.sub(pattern, replace_match, text)


def _replace_variables_recursive(
    obj: Any, task_data: Optional["ExecutionRequest"]
) -> Any:
    """
    Recursively process an object and replace placeholders in all string values.

    Args:
        obj: The object to process (dict, list, string, or other)
        task_data: Data source for replacement values (ExecutionRequest object).

    Returns:
        The processed object with all string placeholders replaced
    """
    if isinstance(obj, dict):
        return {
            key: _replace_variables_recursive(value, task_data)
            for key, value in obj.items()
        }
    elif isinstance(obj, list):
        return [_replace_variables_recursive(item, task_data) for item in obj]
    elif isinstance(obj, str):
        return _replace_placeholders_in_string(obj, task_data)
    else:
        # For other types (int, float, bool, None, etc.), return as-is
        return obj


def replace_mcp_server_variables(
    mcp_servers: Optional[Any], task_data: Optional["ExecutionRequest"]
) -> Optional[Any]:
    """
    Replace ${{path}} placeholders in MCP servers configuration with values from task_data.

    This function recursively traverses the mcp_servers configuration and replaces
    all ${{path.to.value}} format placeholders with actual values from task_data.

    Args:
        mcp_servers: The MCP servers configuration dictionary
        task_data: Data source for replacement values. Must be an ExecutionRequest object.
                   Supports nested access like "user.name" -> task_data.user["name"]
                   or "bot.0.name" -> task_data.bot[0]["name"].

    Returns:
        A new dictionary with all placeholders replaced. If a path doesn't exist
        in task_data, the original placeholder is preserved.

    Examples:
        >>> from shared.models.execution import ExecutionRequest
        >>> mcp_servers = {
        ...     "server1": {
        ...         "url": "https://api.example.com/${{user.git_login}}",
        ...         "headers": {
        ...             "Authorization": "Bearer ${{user.git_token}}",
        ...             "X-User": "${{user.name}}"
        ...         }
        ...     }
        ... }
        >>> task_data = ExecutionRequest(
        ...     user={
        ...         "name": "zhangsan",
        ...         "git_login": "zhangsan",
        ...         "git_token": "token123"
        ...     }
        ... )
        >>> result = replace_mcp_server_variables(mcp_servers, task_data)
        >>> result["server1"]["url"]
        'https://api.example.com/zhangsan'
        >>> result["server1"]["headers"]["Authorization"]
        'Bearer token123'

    Supported placeholder paths:
        - Simple: ${{git_repo}}, ${{branch_name}}
        - Nested: ${{user.name}}, ${{user.git_token}}
        - Deep nested: ${{bot.0.agent_config.env.api_key}} (for list access, use index)
    """
    if not mcp_servers:
        logger.debug("Empty mcp_servers provided, returning mcp_servers")
        return mcp_servers

    if not task_data:
        logger.debug("Empty task_data provided, returning mcp_servers unchanged")
        return mcp_servers

    try:
        result = _replace_variables_recursive(mcp_servers, task_data)
        logger.info("Successfully replaced variables in MCP servers configuration")
        return result
    except Exception as e:
        logger.warning(f"Failed to replace variables in MCP servers: {str(e)}")
        return mcp_servers

File: executor/src/mcp_utils.rs
// SPDX-FileCopyrightText: 2026 Weibo, Inc.
//
// SPDX-License-Identifier: Apache-2.0

use serde_json::Value;

use crate::protocol::ExecutionRequest;

pub fn extract_mcp_servers_config(config: &Value) -> Option<&Value> {
    let object = config.as_object()?;
    let mut candidates = Vec::new();

    for outer_key in ["mcpServers", "mcp_servers"] {
        let Some(value) = object
            .get(outer_key)
            .filter(|value| is_non_empty_object(value))
        else {
            continue;
        };
        let mut found_double = false;
        for inner_key in ["mcpServers", "mcp_servers"] {
            if let Some(nested) = value
                .get(inner_key)
                .filter(|nested| is_non_empty_object(nested))
            {
                candidates.push(McpConfigCandidate::double(outer_key, inner_key, nested));
                found_double = true;
            }
        }
        if !found_double {
            candidates.push(McpConfigCandidate::single(outer_key, value));
        }
    }

    select_mcp_config_candidate(&candidates).or_else(|| object.get("mcp_servers"))
}

pub fn replace_mcp_server_variables(
    mcp_servers: &Value,
    task_data: Option<&ExecutionRequest>,
) -> Value {
    if is_empty_json(mcp_servers) {
        return mcp_servers.clone();
    }
    let Some(task_data) = task_data else {
        return mcp_servers.clone();
    };
    let source = task_data.variable_context();
    replace_variables_recursive(mcp_servers, Some(&source))
}

pub fn replace_variables_recursive(value: &Value, source: Option<&Value>) -> Value {
    match value {
        Value::Object(object) => Value::Object(
            object
                .iter()
                .map(|(key, value)| (key.clone(), replace_variables_recursive(value, source)))
                .collect(),
        ),
        Value::Array(values) => Value::Array(
            values
                .iter()
                .map(|value| replace_variables_recursive(value, source))
                .collect(),
        ),
        Value::String(text) => Value::String(replace_placeholders_in_string(text, source)),
        value => value.clone(),
    }
}

pub fn replace_placeholders_in_string(text: &str, source: Option<&Value>) -> String {
    let mut output = String::with_capacity(text.len());
    let mut cursor = 0;

    while let Some(start_offset) = text[cursor..].find("${{") {
        let start = cursor + start_offset;
        output.push_str(&text[cursor..start]);
        let path_start = start + 3;
        let Some(end_offset) = text[path_start..].find("}}") else {
            output.push_str(&text[start..]);
            return output;
        };
        let end = path_start + end_offset;
        let placeholder_end = end + 2;
        let placeholder = &text[start..placeholder_end];
        let path = text[path_start..end].trim();

        if let Some(value) = get_nested_value(source, path).filter(|value| !value.is_null()) {
            output.push_str(&value_to_placeholder_text(value));
        } else {
            output.push_str(placeholder);
        }
        cursor = placeholder_end;
    }

    output.push_str(&text[cursor..]);
    output
}

pub fn get_nested_value<'a>(source: Option<&'a Value>, path: &str) -> Option<&'a Value> {
    let mut current = source?;
    let path = path.trim();
    if path.is_empty() {
        return None;
    }

    for key in path.split('.') {
        current = match current {
            Value::Object(object) => object.get(key)?,
            Value::Array(values) => {
                let index = key.parse::<usize>().ok()?;
                values.get(index)?
            }
            _ => return None,
        };
    }

    Some(current)
}

fn value_to_placeholder_text(value: &Value) -> String {
    match value {
        Value::String(text) => text.clone(),
        Value::Number(number) => number.to_string(),
        Value::Bool(value) => value.to_string(),
        value => value.to_string(),
    }
}

fn select_mcp_config_candidate<'a>(candidates: &[McpConfigCandidate<'a>]) -> Option<&'a Value> {
    for (outer_key, inner_key) in [
        ("mcpServers", Some("mcpServers")),
        ("mcpServers", Some("mcp_servers")),
        ("mcp_servers", Some("mcp_servers")),
        ("mcp_servers", Some("mcpServers")),
        ("mcpServers", None),
        ("mcp_servers", None),
    ] {
        if let Some(candidate) = candidates
            .iter()
            .find(|candidate| candidate.outer_key == outer_key && candidate.inner_key == inner_key)
        {
            return Some(candidate.config);
        }
    }
    None
}

fn is_non_empty_object(value: &Value) -> bool {
    value.as_object().is_some_and(|object| !object.is_empty())
}

fn is_empty_json(value: &Value) -> bool {
    match value {
        Value::Null => true,
        Value::Array(values) => values.is_empty(),
        Value::Object(object) => object.is_empty(),
        _ => false,
    }
}

#[derive(Debug)]
struct McpConfigCandidate<'a> {
    outer_key: &'static str,
    inner_key: Option<&'static str>,
    config: &'a Value,
}

impl<'a> McpConfigCandidate<'a> {
    fn double(outer_key: &'static str, inner_key: &'static str, config: &'a Value) -> Self {
        Self {
            outer_key,
            inner_key: Some(inner_key),
            config,
        }
    }

    fn single(outer_key: &'static str, config: &'a Value) -> Self {
        Self {
            outer_key,
            inner_key: None,
            config,
        }
    }
}
